src/trel_dnssd/trel_dnssd.hpp

```diff
@@ -0,0 +1,211 @@
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for TREL DNS-SD over mDNS.
+ */
+
+#ifndef OTBR_AGENT_TREL_DNSSD_HPP_
+#define OTBR_AGENT_TREL_DNSSD_HPP_
+
+#if OTBR_ENABLE_TREL
+
+#include <assert.h>
+#include <utility>
+
+#include <openthread/instance.h>
+
+#include "common/types.hpp"
+#include "mdns/mdns.hpp"
+#include "ncp/ncp_openthread.hpp"
+
+namespace otbr {
+
+namespace TrelDnssd {
+
+/**
+ * @addtogroup border-router-trel-dnssd
+ *
+ * @brief
+ *   This module includes definition for TREL DNS-SD over mDNS.
+ *
+ * @{
+ */
+
+class TrelDnssd
+{
+public:
+    /**
+     * This constructor initializes the TrelDnssd instance.
+     *
+     * @param[in] aNcp        A reference to the OpenThread Controller instance.
+     * @param[in] aPublisher  A reference to the mDNS Publisher.
+     *
+     */
+    explicit TrelDnssd(Ncp::ControllerOpenThread &aNcp, Mdns::Publisher &aPublisher)
+        : mPublisher(aPublisher)
+        , mNcp(aNcp)
+    {
+        assert(sTrelDnssd == nullptr);
+
+        sTrelDnssd = this;
+    }
+
+    /**
+     * This method gets the singleton TrelDnssd instance.
+     *
+     * @returns The singleton TrelDnssd instance.
+     *
+     */
+    static TrelDnssd &GetInstance(void) { return *sTrelDnssd; }
+
+    /**
+     * This method initializes the TrelDnssd instance.
+     *
+     * @param[in] aTrelNetif  The network interface for discovering TREL peers.
+     *
+     */
+    void Initialize(std::string aTrelNetif);
+
+    /**
+     * This method starts browsing for TREL peers.
+     *
+     */
+    void StartBrowse(void);
+
+    /**
+     * This method stops browsing for TREL peers.
+     *
+     */
+    void StopBrowse(void);
+
+    /**
+     * This method registers the TREL service to DNS-SD.
+     *
+     * @param[in] aPort         The UDP port of TREL service.
+     * @param[in] aTxtData      The TXT data of TREL service.
+     * @param[in] aTxtLength    The TXT length of TREL service.
+     *
+     */
+    void RegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength);
```

Comment:
is it better to rename it to `PublishService`? As we are publishing the TREL service to be discovered by peers.

Reply:
We already have `PublishTrelService` which do the actual publishing. 

`RegisterService` is called by `trelDnssdRegisterService`, so should better keep `RegisterService`.